File: backend/app/services/imports.py
"""Spreadsheet parsing for product and retailer imports.

Mirrors frontend/src/onboarding/imports.ts: same header aliases, limits and error wording, so a file that
passes client-side validation passes here too, and vice versa.
"""

import csv
import io
import re
from dataclasses import dataclass, field
from decimal import Decimal, InvalidOperation

from openpyxl import load_workbook

from app.core.config import settings
from app.core.security import normalise_phone

EMAIL_RE = re.compile(r"^[^\s@]+@[^\s@]+\.[^\s@]{2,}$")


class ImportError_(Exception):
    """Raised for whole-file problems (wrong type, missing columns) or when rows fail validation."""


def _header_key(value: str) -> str:
    return re.sub(r"[^a-z0-9]", "", value.lower())


def _column(headers: list[str], aliases: list[str]) -> int:
    keys = {_header_key(a) for a in aliases}
    for i, h in enumerate(headers):
        if _header_key(h) in keys:
            return i
    return -1


def _cell(row: list[str], index: int) -> str:
    return row[index] if 0 <= index < len(row) else ""


def _numeric(value: str, fallback: Decimal | None = None) -> Decimal | None:
    if value == "":
        return fallback
    cleaned = re.sub(r"NGN|₦|,", "", value, flags=re.I).strip()
    try:
        return Decimal(cleaned) if cleaned else None
    except InvalidOperation:
        return None


def read_spreadsheet(file_name: str, content: bytes) -> list[list[str]]:
    if not content:
        raise ImportError_("This file is empty. Add a header row and at least one record.")
    if len(content) > settings.import_max_bytes:
        raise ImportError_("This file is a little too large. Please use a file under 5 MB.")
    ext = file_name.rsplit(".", 1)[-1].lower() if "." in file_name else ""
    if ext not in {"csv", "xlsx"}:
        raise ImportError_(
            "Please choose a CSV or Excel (.xlsx) file. Older .xls files can be saved as .xlsx in Excel."
        )

    rows: list[list[str]]
    if ext == "csv":
        text = content.decode("utf-8-sig", errors="replace")
        try:
            rows = list(csv.reader(io.StringIO(text)))
        except csv.Error as exc:
            raise ImportError_(
                "Some quotation marks in this CSV do not match. Please check the file and try again."
            ) from exc
    else:
        try:
            wb = load_workbook(io.BytesIO(content), read_only=True, data_only=True)
            ws = wb.worksheets[0]
            rows = [["" if c is None else str(c) for c in r] for r in ws.iter_rows(values_only=True)]
        except Exception as exc:
            raise ImportError_(
                "We couldn't read this workbook. Please check it isn't password protected, or save it as CSV."
            ) from exc

    rows = [[c.strip() for c in row] for row in rows]
    rows = [row for row in rows if any(row)]
    if len(rows) < 2:
        raise ImportError_("Add column headers and at least one record to your file.")
    if len(rows) > settings.import_max_rows + 1:
        raise ImportError_("Please import up to 5,000 records at a time.")
    return rows


def _raise_row_errors(errors: list[str]) -> None:
    if errors:
        head = " ".join(errors[:3])
        tail = f" Plus {len(errors) - 3} more issues." if len(errors) > 3 else ""
        raise ImportError_(head + tail)


@dataclass
class ProductRow:
    name: str
    sku: str
    price: Decimal
    stock: int
    unit: str | None = None


@dataclass
class RetailerRow:
    name: str
    email: str | None
    phone: str | None
    market: str | None = None
    aliases: list[str] = field(default_factory=list)


def parse_products(rows: list[list[str]]) -> list[ProductRow]:
    headers, *data = rows
    name_col = _column(headers, ["product", "product name", "name", "item", "item name"])
    price_col = _column(headers, ["price", "price ngn", "unit price", "unit price ngn", "selling price", "amount"])
    sku_col = _column(headers, ["sku", "product code", "item code", "code"])
    stock_col = _column(
        headers, ["stock", "quantity", "qty", "stock quantity", "stock level", "quantity available", "stock on hand"]
    )
    unit_col = _column(headers, ["unit", "pack", "pack size", "unit size"])
    if name_col < 0 or price_col < 0:
        raise ImportError_(
            'Your file needs "Product name" and "Price" columns. SKU and Stock are optional. '
            "Our template has the right headings."
        )

    errors: list[str] = []
    seen: set[str] = set()
    products: list[ProductRow] = []
    for index, row in enumerate(data):
        line = index + 2
        name = _cell(row, name_col)
        price = _numeric(_cell(row, price_col))
        stock = _numeric(_cell(row, stock_col), Decimal(0))
        sku = (_cell(row, sku_col) or f"LEDA{index + 1:04d}").upper()
        if not name:
            errors.append(f"Row {line}: product name is missing.")
        if price is None or price < 0:
            errors.append(f"Row {line}: enter a valid, nonnegative price.")
        if stock is None or stock < 0 or stock != stock.to_integral_value():
            errors.append(f"Row {line}: stock must be a nonnegative whole number.")
        if sku in seen:
            errors.append(f"Row {line}: SKU {sku} is repeated.")
        seen.add(sku)
        products.append(
            ProductRow(
                name=name,
                sku=sku,
                price=price or Decimal(0),
                stock=int(stock or 0),
                unit=_cell(row, unit_col) or None,
            )
        )
    _raise_row_errors(errors)
    return products


def parse_retailers(rows: list[list[str]]) -> list[RetailerRow]:
    headers, *data = rows
    name_col = _column(headers, ["retailer name", "retailer", "business name", "name", "customer", "customer name"])
    email_col = _column(headers, ["email", "email address", "retailer email"])
    phone_col = _column(
        headers,
        [
            "phone",
            "phone number",
            "phone no",
            "mobile",
            "mobile number",
            "whatsapp",
            "whatsapp number",
            "telephone",
            "contact number",
        ],
    )
    market_col = _column(headers, ["market", "location", "area", "address"])
    if name_col < 0 or (email_col < 0 and phone_col < 0):
        raise ImportError_(
            'Your file needs a "Retailer name" column and at least one contact column: "Email" or "Phone".'
        )

    errors: list[str] = []
    contacts: set[str] = set()
    retailers: list[RetailerRow] = []
    for index, row in enumerate(data):
        line = index + 2
        name = _cell(row, name_col)
        email = _cell(row, email_col).lower()
        raw_phone = _cell(row, phone_col)
        phone = normalise_phone(raw_phone) if raw_phone else None
        if not name:
            errors.append(f"Row {line}: retailer name is missing.")
        if not email and not raw_phone:
            errors.append(f"Row {line}: add an email or phone number.")
        if email and not EMAIL_RE.match(email):
            errors.append(f"Row {line}: the email is not valid.")
        if raw_phone and not phone:
            errors.append(f"Row {line}: use a valid Nigerian phone number.")
        key = email or phone
        if key and key in contacts:
            errors.append(f"Row {line}: this contact appears more than once.")
        if key:
            contacts.add(key)
        retailers.append(
            RetailerRow(name=name, email=email or None, phone=phone, market=_cell(row, market_col) or None)
        )
    _raise_row_errors(errors)
    return retailers

File: frontend/src/onboarding/imports.ts
import Papa from "papaparse";
import { readSheet } from "read-excel-file/browser";
import { downloadFile, normalisePhone, validEmail, type Product, type Retailer } from "./model";

const MAX_ROWS = 5000;
const MAX_BYTES = 5 * 1024 * 1024;
const headerKey = (value: string) => value.toLowerCase().replace(/[^a-z0-9]/g, "");

export async function readSpreadsheet(file: File): Promise<string[][]> {
  if (!file.size) throw new Error("This file is empty. Add a header row and at least one record.");
  if (file.size > MAX_BYTES) throw new Error("This file is a little too large. Please use a file under 5 MB.");
  const extension = file.name.split(".").pop()?.toLowerCase();
  if (extension !== "csv" && extension !== "xlsx") {
    throw new Error("Please choose a CSV or Excel (.xlsx) file. Older .xls files can be saved as .xlsx in Excel.");
  }

  let rows: string[][];
  if (extension === "csv") {
    const result = Papa.parse<string[]>(await file.text(), { skipEmptyLines: "greedy", preview: MAX_ROWS + 2 });
    if (result.errors.some(error => error.type === "Quotes")) {
      throw new Error("Some quotation marks in this CSV do not match. Please check the file and try again.");
    }
    rows = result.data;
  } else {
    try {
      const data = await readSheet(file, 1);
      rows = data.map(row => row.map(cell => String(cell ?? "")));
    } catch {
      throw new Error("We couldn't read this workbook. Please check it isn't password protected, or save it as CSV.");
    }
  }

  rows = rows.map(row => row.map(cell => cell.trim())).filter(row => row.some(Boolean));
  if (rows.length < 2) throw new Error("Add column headers and at least one record to your file.");
  if (rows.length > MAX_ROWS + 1) throw new Error("Please import up to 5,000 records at a time.");
  return rows;
}

function column(headers: string[], aliases: string[]) {
  return headers.findIndex(header => aliases.map(headerKey).includes(headerKey(header)));
}

function numeric(value: string | undefined, fallback?: number) {
  if (value === undefined || value === "") return fallback ?? NaN;
  const cleaned = value.replace(/NGN|\u20a6|,/gi, "").trim();
  return cleaned ? Number(cleaned) : NaN;
}

export function parseProducts(rows: string[][]): Product[] {
  const [headers, ...data] = rows;
  const nameColumn = column(headers, ["product", "product name", "name", "item", "item name"]);
  const priceColumn = column(headers, ["price", "price ngn", "unit price", "unit price ngn", "selling price", "amount"]);
  const skuColumn = column(headers, ["sku", "product code", "item code", "code"]);
  const stockColumn = column(headers, ["stock", "quantity", "qty", "stock quantity", "stock level", "quantity available", "stock on hand"]);
  if (nameColumn < 0 || priceColumn < 0) {
    throw new Error('Your file needs "Product name" and "Price" columns. SKU and Stock are optional. Our template has the right headings.');
  }

  const errors: string[] = [];
  const skus = new Set<string>();
  const products = data.map((row, index) => {
    const name = row[nameColumn] ?? "";
    const price = numeric(row[priceColumn]);
    const stock = numeric(row[stockColumn], 0);
    const sku = (row[skuColumn] || `LEDA${String(index + 1).padStart(4, "0")}`).toUpperCase();
    if (!name) errors.push(`Row ${index + 2}: product name is missing.`);
    if (!Number.isFinite(price) || price < 0) errors.push(`Row ${index + 2}: enter a valid, nonnegative price.`);
    if (!Number.isSafeInteger(stock) || stock < 0) errors.push(`Row ${index + 2}: stock must be a nonnegative whole number.`);
    if (skus.has(sku)) errors.push(`Row ${index + 2}: SKU ${sku} is repeated.`);
    skus.add(sku);
    return { name, price, stock, sku };
  });
  if (errors.length) throw new Error(errors.slice(0, 3).join(" ") + (errors.length > 3 ? ` Plus ${errors.length - 3} more issues.` : ""));
  return products;
}

export function parseRetailers(rows: string[][]): Retailer[] {
  const [headers, ...data] = rows;
  const nameColumn = column(headers, ["retailer name", "retailer", "business name", "name", "customer", "customer name"]);
  const emailColumn = column(headers, ["email", "email address", "retailer email"]);
  const phoneColumn = column(headers, ["phone", "phone number", "phone no", "mobile", "mobile number", "whatsapp", "whatsapp number", "telephone", "contact number"]);
  if (nameColumn < 0 || (emailColumn < 0 && phoneColumn < 0)) {
    throw new Error('Your file needs a "Retailer name" column and at least one contact column: "Email" or "Phone".');
  }
  const errors: string[] = [];
  const contacts = new Set<string>();
  const retailers = data.map((row, index) => {
    const name = row[nameColumn] ?? "";
    const email = (row[emailColumn] ?? "").toLowerCase();
    const rawPhone = row[phoneColumn] ?? "";
    const phone = rawPhone ? normalisePhone(rawPhone) ?? "" : "";
    if (!name) errors.push(`Row ${index + 2}: retailer name is missing.`);
    if (!email && !rawPhone) errors.push(`Row ${index + 2}: add an email or phone number.`);
    if (email && !validEmail(email)) errors.push(`Row ${index + 2}: the email is not valid.`);
    if (rawPhone && !phone) errors.push(`Row ${index + 2}: use a valid Nigerian phone number.`);
    const key = email || phone;
    if (key && contacts.has(key)) errors.push(`Row ${index + 2}: this contact appears more than once.`);
    if (key) contacts.add(key);
    // Preview references cannot be mistaken for usable bank account numbers.
    return { name, email, phone, accountReference: `PREVIEW ${String(index + 1).padStart(4, "0")}` };
  });
  if (errors.length) throw new Error(errors.slice(0, 3).join(" ") + (errors.length > 3 ? ` Plus ${errors.length - 3} more issues.` : ""));
  return retailers;
}

export const SAMPLE_PRODUCTS = [
  ["Product name", "SKU", "Price", "Stock"],
  ["Royal Stallion Rice 50kg", "RSR50", "78500", "120"],
  ["Mama Gold Rice 50kg", "MGR50", "77200", "80"],
  ["Kings Vegetable Oil 25L", "KVO25", "96500", "65"],
];

export const SAMPLE_RETAILERS = [
  ["Retailer name", "Email", "Phone"],
  ["Kike Stores", "kike@example.com", "08031234567"],
  ["Okafor Provisions", "okafor@example.com", "08052345678"],
  ["Amina Food Mart", "amina@example.com", "07063456789"],
];

export function downloadTemplate(kind: "products" | "retailers") {
  const data = kind === "products" ? SAMPLE_PRODUCTS : SAMPLE_RETAILERS;
  const csv = Papa.unparse(data, { escapeFormulae: true });
  downloadFile(`\uFEFF${csv}`, `leda-${kind}-template.csv`, "text/csv;charset=utf-8;");
}